uportal-war/src/main/java/org/jasig/portal/security/provider/AnyUnblockedGrantPermissionPolicy.java

```diff
@@ -181,11 +181,44 @@ public boolean doesPrincipalHavePermission(
 
     }
 
+    /**
+     * Allows an outside actor to force this policy to evaluate and cache an
+     * authorization decision.  Permissions checking can be expensive;  a
+     * well-primed cache can make the task perform better.  This method will
+     * create the cache entry whether it exists already or not, forcibly
+     * resetting the TTL.
+     *
+     * @since uPortal 4.3
+     */
+    public void loadInCache(IAuthorizationService service,
+            IAuthorizationPrincipal principal, IPermissionOwner owner,
+            IPermissionActivity activity, IPermissionTarget target) {
+
+        final Set<IGroupMember> seenGroups = new HashSet<>();
+        final CacheTuple cacheTuple = new CacheTuple(
+                principal.getPrincipalString(),
+                owner.getFname(),
+                activity.getFname(),
+                target.getKey());
+        final boolean answer = hasUnblockedPathToGrant(service, principal, owner, activity, target, seenGroups);
+        Element element = new Element(cacheTuple, answer);
+        hasUnblockedGrantCache.put(element);
+
+    }
+
     private boolean hasUnblockedPathToGrantWithCache(IAuthorizationService service,
             IAuthorizationPrincipal principal, IPermissionOwner owner,
             IPermissionActivity activity, IPermissionTarget target,
             Set<IGroupMember> seenGroups) throws GroupsException {
 
+        /*
+         * We're only going to use the caches for groups;  caching these items
+         * for users would be (1) too large and (2) redundant with other caches.
+         */
```

Comment:
This part may have been a mistake.  I'm testing to see if it accounts for the extra (bad) 500ms I'm seeing in the microbenchmarking.